miner/Miner.hs

```diff
@@ -0,0 +1,285 @@
+
+--------------------------------------------------------------------------------
+-- CLI
+
+type ResultMap = HashMap (T2 ChainId BlockHeight) BlockHeader
+
+data Env = Env
+    { work :: TMVar BlockHeader
+    , results :: TVar ResultMap
+    , cores :: Word16
+    , version :: ChainwebVersion
+    , port :: Int
+    }
+
+pEnv :: TMVar BlockHeader -> TVar ResultMap -> Parser Env
+pEnv tbh thm = Env tbh thm <$> pCores <*> pVersion <*> pPort
+
+pCores :: Parser Word16
+pCores = option auto
+    (long "cores" <> metavar "COUNT" <> value 1
+     <> help "Number of CPU cores to use (default: 1)")
+
+pVersion :: Parser ChainwebVersion
+pVersion = option cver
+    (long "version" <> metavar "VERSION"
+     <> value Testnet01
+     <> help "Chainweb Network Version (default: testnet01)")
+  where
+    cver :: ReadM ChainwebVersion
+    cver = eitherReader $ \s ->
+        note "Illegal ChainwebVersion" . chainwebVersionFromText $ T.pack s
+
+pPort :: Parser Int
+pPort = option auto
+    (long "port" <> metavar "PORT" <> value 8081
+     <> help "Port on which to run the miner (default: 8081)")
+
+--------------------------------------------------------------------------------
+-- Work
+
+main :: IO ()
+main = do
+    env <- (opts <$> newEmptyTMVarIO <*> newTVarIO mempty) >>= execParser
+    miner <- async $ mining env
+    W.run (port env) $ app env
+    wait miner
+  where
+    opts :: TMVar BlockHeader -> TVar ResultMap -> ParserInfo Env
+    opts tbh thm = info (pEnv tbh thm <**> helper)
+        (fullDesc <> progDesc "The Official Chainweb Mining Client")
+
+-- | Submit a new `BlockHeader` to mine (i.e. to determine a valid `Nonce`).
+--
+submit :: Env -> BlockHeader -> IO ()
+submit (work -> w) bh = atomically $
+    isEmptyTMVar w >>= bool (void $ swapTMVar w bh) (putTMVar w bh)
+
+-- | For some `ChainId` and `BlockHeight`, have we mined a result?
+--
+poll :: Env -> ChainId -> BlockHeight -> IO (Maybe BlockHeader)
+poll (results -> tm) cid h = HM.lookup (T2 cid h) <$> readTVarIO tm
+
+-- | Cease mining until another `submit` call is made.
+--
+-- halt :: IO ()
+-- halt = undefined
+
+-- | A supervisor thread that listens for new work and supervises mining threads.
+--
+mining :: Env -> IO ()
+mining e = do
+    bh <- atomically . takeTMVar $ work e
+    race newWork (go bh) >>= traverse_ miningSuccess
+    mining e
+  where
+    -- | Wait for new work to come in from a `poll` call. `readTMVar` has the
+    -- effect of "waiting patiently" and will not cook the CPU.
+    newWork :: IO ()
+    newWork = void . atomically . readTMVar $ work e
+
+    -- | If the `go` call won the `race`, this functions saves the result of
+    -- that successful mining. If `newWork` won the race instead, then the `go`
+    -- call is automatically cancelled.
+    miningSuccess :: BlockHeader -> IO ()
+    miningSuccess new = do
+        let key = T2 (_blockChainId new) (_blockHeight new)
+        atomically $ modifyTVar' (results e) (HM.insert key new)
```

Comment:
shouldn't this call `PUT cut` on consensus for the newly minded block? Or are we planning to add another component around this that is sitting in-between this miner and consensus?

Reply:
I guess a more general question is: Mining is generally on a per-cut level. Since this component is aware of `BlockHeaders`, chains, and block heights (as opposed to a low-level miner that only finds a nonce), shouldn't it be rather work on a cut level?

Essentially, it would listen for new cuts (via submit) mine a new cut (by picking a chain and computing a block), and when successful submit the cut to consensus via `PUT cut`.